### ISSUE TITLE
Native systemd units to make use of makefs and growfs capabilities without using /etc/fstab

### ISSUE THREAD

**Issue by frittentheke** [systemd/systemd#10014]
**Is your feature request related to a problem? Please describe.**
With Systemd 236 the options _x-systemd.makefs_ and _x-systemd.growfs_ where introduced (https://lists.freedesktop.org/archives/systemd-devel/2017-December/039996.html) to allow automagic filesystem creation and growing of filesystems.

Unfortunately there seems to be no native way of creating mount units that will then trigger those units to be created as well. There was some discussion about creating a whole framework to do that - https://github.com/systemd/systemd/pull/7237#discussion_r154321006

Using very modern / stripped down distributions, e.g. Redhat Container-Linux, which only use systemd.mount units and spare themselves writing any _/etc/fstab_ this unfortunately renders using the those systemd features very difficult to use.

**Describe the solution you'd like**
Certainly the solution would be to have some options for the mount unit to makefs and/or growfs, just like with those in fstab to auto-create the corresponding makefs/growfs units.

**Describe alternatives you've considered**
Providing skeleton units / examples of for systemd-makefs@.service and systemd-growfs@.service units would be helpful as well, to allow manually (without a simple mount unit option) to only use systemd to mange filesystems and their mount - no scripts being triggered - which is what has to be done / seems to only way currently.

**Comment by poettering** [comment]
Yeah, we should really ship non-generated versions of systemd-makefs@.service and systemd-growfs@.service indeed. Right now fstab-generator creates them (which is nice as they can be adjusted dynamically to the device), but much the same as for systemd-fsck@.service we really should have a regular template unit for these too.

**Comment by akira-kurogane** [comment]
Note - the cross-mention to "mount namespace cleanups # 10012" above is not genuine. It looks like it was auto-created from a typo in a first draft of https://github.com/systemd/systemd/pull/10012#issuecomment-422261948, but these two issues are unrelated.

**Comment by frittentheke** [comment]
> Yeah, we should really ship non-generated versions of systemd-makefs@.service and systemd-growfs@.service indeed. Right now fstab-generator creates them (which is nice as they can be adjusted dynamically to the device), but much the same as for systemd-fsck@.service we really should have a regular template unit for these too.

@poettering any chance the units you mentioned could be added?

**Comment by frittentheke** [comment]
@poettering  ... could you kindly help me understand how far this issue has potentially be resolved?
I suppose in https://github.com/systemd/systemd/issues/20788 you added the growfs units, so what's left to do are those for makefs?

**Comment by poettering** [comment]
yeah, growfs is handled now. makefs is missing. makefs is harder since it typically needs two args: device node *and* fstype. We so far only have support for a single "instance" id whith instantiated services, which hence makes the mapping harder.

that said, we probably could make an instantiation of `systemd-makefs@<node>:<fstype>.service`  work or so.

or we could make it smart, so that if you instantiate it just as `systemd-makefs@.service` it would pick up the fstype of the rootfs (or /var/?) and then use that same fstype for whatever it is invoked. Under the assumption that most deployments probably go for the same fstype for all fs they intend to connect to it that should be a reasonably sane default I think

**Comment by Jamesits** [comment]
@poettering Is it possible to make `systemd-makefs@.service` read the synthetic `fstab` (incl. `.mount` units) for filesystem types? End user can still disable auto mounting of the file system by other means, e.g. not attaching the `.mount` unit to the `*-fs.target` as a dependency.

**Comment by poettering** [comment]
there is no synthetic fstab. and no, i doubt systemd-makefs should try to reimplement pid1's complete unit file parsing. this is not how that is supposed to work.

**Comment by septatrix** [comment]
Isn't makefs now mostly covered by sd-repart?

**Comment by frittentheke** [comment]
@poettering may I kindly ask for an update on the makefs capabilities of systemd?
